rust/automerge/src/op_observer/vec_observer.rs

```diff
@@ -0,0 +1,561 @@
+#![allow(dead_code)]
+
+use core::fmt::Debug;
+
+use crate::{marks::Mark, ObjId, OpObserver, Prop, ReadDoc, ScalarValue, Value};
+
+use crate::sequence_tree::SequenceTree;
+
+use crate::op_observer::BranchableObserver;
+use crate::op_observer::{Patch, PatchAction};
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum TextRepresentation {
+    Array,
+    String,
+}
+
+impl TextRepresentation {
+    pub fn is_array(&self) -> bool {
+        matches!(self, TextRepresentation::Array)
+    }
+
+    pub fn is_string(&self) -> bool {
+        matches!(self, TextRepresentation::String)
+    }
+}
+
+impl std::default::Default for TextRepresentation {
+    fn default() -> Self {
+        TextRepresentation::Array
+    }
+}
+
+pub(crate) trait TextIndex {
+    type Item: Debug + PartialEq + Clone;
+    type Iter<'a>: Iterator<Item = Self::Item>;
+
+    fn chars(text: &str) -> Self::Iter<'_>;
+}
+
+#[derive(Debug, Clone, Default)]
+struct VecOpObserverInner<T: TextIndex> {
+    pub(crate) patches: Vec<Patch<T::Item>>,
+    pub(crate) text_rep: TextRepresentation,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct VecOpObserver(VecOpObserverInner<Utf8TextIndex>);
+
+#[derive(Debug, Clone, Default)]
+pub struct VecOpObserver16(VecOpObserverInner<Utf16TextIndex>);
```

Comment:
We should document that this is the same as `VecOpObserver` but text indices will use utf-16 offsets rather than `utf-8`.